src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkBaseExternalContext.java

```diff
@@ -1349,24 +1349,42 @@ during the analysis phase and thus cannot depends on a target result (the \
             },
             doc =
                 "<code>string</code>, <code>Label</code> or <code>path</code> from which to create"
-                    + " a path from.")
+                    + " a path from."),
+        @Param(
+            name = "watch",
+            defaultValue = "'auto'",
+            positional = false,
+            named = true,
+            doc =
+                """
+                Whether to <a href="#watch">watch</a> the path. Can be the string 'yes', 'no', \
+                or 'auto'. Passing 'yes' is equivalent to immediately invoking the \
+                <a href="#watch"><code>watch()</code></a> method; passing 'no' does not \
+                attempt to watch the path; passing 'auto' will only attempt to watch the \
+                path when it is a <code>Label</code> and legal to do so (see <code>watch()</code> \
+                docs for more information).
+                """)
       })
-  public StarlarkPath path(Object path) throws EvalException, InterruptedException {
-    return getPath("path()", path);
+  public StarlarkPath path(Object path, String watch)
+      throws EvalException, InterruptedException, RepositoryFunctionException {
+    var shouldWatch = ShouldWatch.fromString(watch);
+    var result =
+        switch (path) {
+          case String s -> new StarlarkPath(this, workingDirectory.getRelative(s));
+          case Label label -> getPathFromLabel(label, shouldWatch);
+          case StarlarkPath starlarkPath -> starlarkPath;
+            // This can never happen because we check it in the Starlark interpreter.
+          default -> throw new IllegalArgumentException("expected string or label for path");
+        };
+    if (shouldWatch == ShouldWatch.YES) {
```

Comment:
why not just `maybeWatch(result, shouldWatch)` here?

Reply:
I added a comment, this would be backwards incompatible (caught by one of the new tests).

Reply:
(sorry for the delay)

You're talking about `testPath_watchAuto_starlarkPathAndAbsolutePathNotWatched`, right? I agree that's a change in behavior, but I don't know if it counts as backwards-incompatible. Didn't we do the same behavior change for repo rules?

Plus, we already have the condition where `watch = 'auto'` means that paths outside the workspace aren't watched (https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkBaseExternalContext.java;drc=6052ad65e4185f19944e5138e9b70f6a4eaf9e76;l=1457). Isn't that enough?

Reply:
The PR title was misleading, it adds this parameter to both repo and module context, so we don't have prior art for this function to refer to.

I agree that the change is minor, but I think that it would end up breaking the code in https://github.com/bazelbuild/rules_python/pull/2135/files#diff-ca74c876e894b0bd25547ca666e562281e5c8fb2abc4a51446f410371224aea2 by reintroducing a dependency into the lockfile. Depending on a sibling label, manipulating the path and then obtaining a `StarlarkPath` for it via this method is the kind of existing "pattern" that changing this would break.

Reply:
hmm, I see. Could you add some more details in the PR description about the motivation of this change? It got lost in the commotion.

I _could_ see an API where `auto` (or whatever the default value for `[mr]ctx.path(watch=)` ends up being) means "don't watch". Or just make it default to `no`. Or actually, maybe just make `[mr]ctx.watch` return the path?

Reply:
oh, I remember now -- so `rctx.path(label)` used to watch, and `rctx.path(path)` and `rctx.path(string)` used to _not_ watch.

This PR retains this old `rctx.path(X)` watch behavior (which is based on the type of X, not the actual location of X or whether it's `rctx` or `mctx`). It just allows the caller to be explicit about watching or not by specifying `watch='yes'` or `watch='no'`. The awkward part is that `watch='auto'` means a different thing for `rctx.path()` than `rctx.read()`, `rctx.extract()`, etc.

I wonder if we'd want `rctx.path(X)` to _never_ watch (just use `rctx.watch()` instead). That's obviously backwards-incompatible because some people rely on `rctx.path(label)` watching, but maybe it's worth introducing an incompatible flag flip for? Undecided.

Also, the fact that `rctx.read(label, watch='no')` actually _still_ watches is a bit of a subtle hiccup. I'll need to think about that too...

Reply:
Now that I've thought more about it, I think that I would prefer the following changes:
* `ctx.path` never watches and doesn't have a `watch` parameter. It only ever watched because that was the only way to watch anything at all. It doesn't make much sense that resolving a label to a path is linked to caring about the content at that path. `ctx.watch` can replace it.
* `ctx.read(label, watch = "no")` shouldn't watch.

The latter clearly violates the intent, I would consider changing that behavior a bug fix and don't think it would break anyone.
The former is more impactful to change as it has been the only way to watch for a long time. rules_go would definitely need to change. Luckily `ctx.watch` can be trivially feature detected.

Reply:
I fully agree. Just a bit nervous about not having an incompatible flag for it. (so maybe not so "fully" 😅) Adding an incompatible flag for this one isn't super bad, right? It's just a single `if` in `StarlarkBaseExternalContext` so shouldn't be a lot of maintenance burden. And we can make it no-op in Bazel 9.

Reply:
Yeah, I agree the change for `path` should come with a flag. But since it will mostly require fixing in rulesets, not user projects, we would need to look into that regardless.